### DISCUSSION TITLE
0066 XLS - 66d: XRP Ledger-native Lending Protocol

### DISCUSSION THREAD

**Discussion by Tapanito** [XRPLF/XRPL-Standards#190] [XLS Proposals]
## The earlier version can be found by looking at the edit history.

**Comment by dangell7** [comment]
Personally I would have designed all of this to abide by the current security laws in the US but thats just me.

As these proposals sit, one could invest into a pool, then the managers of that pool could make investment decisions that they don't agree with. This is why we have the current laws set up the way we do in the US. I.E. A prospectus is created, an N-1 is filed, then citizens can invest into the `Fund` and know exactly what they are investing in. Maybe you could view this more as a third party money manager, however a 3PMM still defines clearly what investments the `Fund` may invest in.

These proposals, imo, add complex risk and will likely violate current securities laws but don't let that stop you, I still like the ideas.

**Reply by jazzi-cooper** [reply]
Very aligned with your thinking here. The idea is that the prospectus would be available off-chain, ideally linked in a URI/data field.

Are there other ideas you have to make this more compliant?

**Reply by dangell7** [reply]
If it were me, I would force depositAuth, with DID, verifying that the user has accreditation status. I would also force that the managers have a financial advisor status thought DID. Without them being licensed, them choosing investments for clients is absolutely a violation. 

A lot of this might be off chain stuff, however I would attempt to move as much of this on chain as I can. Or at least make the attempts. I would also move away from renaming things. I would lean more _into_ the current financial instruments than out. For example, the "Pool" is actually what most of us would refer to as a "Fund". Changing the name isn't going to help you in a legal preceding, we obviously know what it is.

**Reply by dangell7** [reply]
To Continue (not necessarily compliance)...

If we do view these as `Funds` then I believe many features of a fund are missing that a lot of managers and investors like to see. I wont go into all of those details but one major component is NAV. Nav is calculated daily because it includes liabilities. I don't see this being taken into consideration other than the lending and the lending fee. There are other fees involved with running a fund. Interest Payable, Taxes Payable and Accrued Liabilities, Front end (Discounts) or Back end (Redemption). If you're doing a no-load fund then you still need a way to charge fees. Managers of a fund need to be paid and this needs to be transparent so that investors are not ripped off by a fund charging 30% to be managed. I would add these directly into the protocol.

**Reply by mvadari** [reply]
That's not how DIDs work. Having a DID doesn't really give you anything, the actual data is stored off-ledger and also can't really be parsed that easily since it involves knowing which VCs you want to trust.
That being said, I do like the idea of having a setting to add deposit auth to the account, so that pool managers can verify this data off-chain before allowing someone to deposit into the pool. I don't think it should be mandatory for all pools, though.

**Reply by dangell7** [reply]
Agree and I think this is why many of us shared concern early about DID. 

They could have been written so licensed third-party verifiers: licensed professionals holding FINRA Series 7 licenses can validated the accreditation status of an individual. Now this status can be verified by any fund without going through the full accreditation status on each users investment.

FINRA could also create an account, validate the Financial Advisors and now all of this exists on chain. (Not the document, but the validation of the document/status).

You can do this all off chain, I'm saying I wouldn't have.

And yes, I wasn't saying mandatory but the fact that this wasn't even mentioned, nor is there any financial disclaimer made me question whether or not these things were being considered.

> I would also think about creating a financial disclaimer. This is an investment vehicle that will likely need to be registered.

**Reply by mvadari** [reply]
> I think this is why many of us shared concern early about DID.

Where were those concerns expressed? I don't see any in the [spec discussion](https://github.com/XRPLF/XRPL-Standards/discussions/100), and I don't recall any voiced elsewhere.

The spec implements the industry-standard [W3C DID spec](https://www.w3.org/TR/did-core/), to avoid reinventing the wheel and [having too many standards](https://xkcd.com/927/) in the industry.

The problem with storing that sort of KYC/financial data directly on-ledger is due to privacy - you want the [verified credential (VC) data](https://www.w3.org/TR/vc-data-model-2.0/) to be stored off-chain because you only want that data for certain use-cases, not for everyone to see on a public blockchain.

Those licensed professionals that you mention could still issue VCs for those details, and the pool manager could verify those details off-chain before approving the deposit auth. If you want it on-chain in some form, it could be encrypted in a memo so the pool manager doesn't need to communicate off-chain with the prospective loan participants. Then you don't need a list of "approved FINRA-licensed addresses" that you somehow have to store on-chain and always keep updated.

**Reply by dangell7** [reply]
[Here](https://github.com/XRPLF/rippled/pull/4636#issuecomment-1659722288) is where we realized you had done it backwards to what we were thinking and would have done. More concerns were voiced internally or on other channels like twitter. 

I never said anything about storing data on the ledger, you just store the attestation.

**Reply by Tapanito** [reply]
Great ideas @dangell7 !  


> If it were me, I would force depositAuth, with DID, verifying that the user has accreditation status. I would also force that the managers have a financial advisor status thought DID. Without them being licensed, them choosing investments for clients is absolutely a violation.

To echo @mvadari it might be quite tricky to add any credential verification logic to the blockchain. The nature of XRPL is that protocols are part of the "core" implementation. Verifying credentials would require coupling `rippled` with some verifiable credential issuer. Though, I like the `depositAuth` idea. Currently, it can be achieved via the `Permissions` field, where a Pool Delegate can explicitly authorise Liquidity Providers and Borrower. (Assuming the Pool Delegate vets the Liqudity Provider and the Borrower off-chain). Do I understand correctly, that you're proposing to enforce that these accounts (Liquidity Providers, Borrower, Pool Delegate) all have associated DIDs with them?

> A lot of this might be off chain stuff, however I would attempt to move as much of this on chain as I can. Or at least make the attempts. I would also move away from renaming things. I would lean more into the current financial instruments than out. For example, the "Pool" is actually what most of us would refer to as a "Fund". Changing the name isn't going to help you in a legal preceding, we obviously know what it is.

Variable naming is hard :) . We chose `Pool` (though perhaps `Vault` might be an even better name) as it indicates a construct that holds a single asset deposited by one or more accounts. It is not necessarily yield-bearing. For example, `Pool` could be used for a multi-party escrow. Lending Protocol is just one application that requires a way to "pool" an asset from multiple accounts.

**Reply by dangell7** [reply]
I think it comes down to who your target market is. If your target market is "crypto" (2 trillion) than Pool works well. However if your target market is Hedge Funds, Financial Advisors and Institutional Finance (23 trillion) than changing the name of something that already exist is just confusing. Also fyi, a multiparty escrow is still called an escrow fund in finance.

Honestly, it really doesn't matter. We're already actively trying to obfuscate the "crypto nomenclature" for everyday users so in this case we would do the same and present them to clients and customers as funds. In the eyes of the law they are funds. Call them whatever you'd like in the backend.

I take back what I said, I would let the creator of the Pool handle the DID/DepositAuth and requirements of their jurisdictional law and not force anything.

**Reply by jazzi-cooper** [reply]
Great points, @dangell7. I agree that we need to broaden our target market beyond just crypto natives. These users will likely use front-end applications that can rename "pools" to better fit their audience. The main audience for this specification is crypto-familiar developers, so I think its reasonable to use established blockchain terminology. For what its worth, we are shifting from the traditional finance model, which is somewhat exclusive and opaque,to a more open, transparent crowd-sourced approach. I think term "pool" aligns well with this new structure.

**Comment by rothexD** [comment]
I feel like a pool should have the option to have some type of uri/data field that it can use to link to external resources, you could put a bunch of usefull stuff behind that like a contacts page of the fund management, page informing about registration (i think lending will almost likely lead to it needing to be registered with the gov in some form) etc

**Reply by jazzi-cooper** [reply]
Yes great catch, this was meant to be included. It is in the pool spec. We will update this spec to reference it.

**Reply by rothexD** [reply]
mb the uri is in pool and this extends pool

i have now read through everything and maybe i missed it but i was looking for something like:

either the pool or this should also have a transfer ownership functionality

pool management may go bust or be bought/sold to others and they will want to ensure all access to mangement gets moved to them

**Reply by Tapanito** [reply]
@rothexD, thank you for your feedback. Indeed, a PoolDelegateURI is part of the `Pool` specification. Would an additional field specific to the `Pool` itself be beneficial?

**Reply by rothexD** [reply]
> @rothexD, thank you for your feedback. Indeed, a PoolDelegateURI is part of the `Pool` specification. Would an additional field specific to the `Pool` itself be beneficial?
> 
> 
> 
> 

I was not expecting it to be in `Pool` as in my mind a `Pool` should be designed so that it can be something generic decentralized with no direct owner.

For example a decentralized tokenized `Pool` of xrp that can be used for flash-loans or similar in the future.

I don't see a need to have an additional uri field

**Reply by Tapanito** [reply]
@rothexD  a similar issue about pool transferabiliy was raised in the `Pool` spec (https://github.com/XRPLF/XRPL-Standards/discussions/192#discussioncomment-9111463) . 

Is decoupling a Pool owner and Delegate and giving them different permissions (including transferring the owner) a similar feature you had in mind?

**Comment by dangell7** [comment]
I like how you included `Failure Conditions` and `State Changes`. 

This makes it very clear what the results of each transaction are.

**Reply by Tapanito** [reply]
Thank you @dangell7 !

**Comment by tequdev** [comment]
It seems to me that a malicious account could take away a user's funds by controlling both the delegator and the borrower, is that correct?
(if `Borrowing Amount > Cover Funds`)

Users have to fully evaluate the delegator.

**Reply by Tapanito** [reply]
Yes, @tequdev, that is correct. This version of the Protocol heavily relies on off-chain processes to prevent your scenario from happening. Liquidity Providers will have to evaluate the Pool Delegates; Pool Delegates will have to assess the Liquidity Providers and the Borrowers. Although assets are issued and repaid on-chain, contractual agreements between all parties will have to be made to enforce the correct behaviour from all parties. 
 The first version of the Protocol is based on creditworthiness. However, future iterations will introduce more sophisticated mechanisms to provide economic incentives to behave correctly. Before we do that, it's essential to get the fundamentals right, which this spec aims to do.

**Comment by outrum** [comment]
I'm particularly interested in how we're planning to tackle potential fraud or manipulation by borrowers or pool delegates. Given the off-chain underwriting and the significant responsibilities pool delegates hold, how are we making sure everything stays above board?

**Reply by jazzi-cooper** [reply]
On other chains, a lot of this happens off-chain between the Application ("marketplace") and the pool delegate. They often have legal contracts in place to ensure pool delegate accountability.

Over time, we would like to bring more of this on-chain, but this is just a v1.

**Comment by ShaneSCalder** [comment]
Are we also thinking of flash loans, leverage or settlement loans? How will this integrate with escrow as some of the basic functionality for lending could be applied or borrowed. Cheers

**Reply by jazzi-cooper** [reply]
Flash loans aren't possible on XRPL today because it doesn't support the necessary atomic transactions. However, upcoming features like XLS-56d (atomic/batch tx's) could potentially change this. As for the use cases you mentioned, they require collateral support, which we haven't included in the first version of our proposal. There is definitely a lot of value to unlock there so we will aim to include in a future version.

**Comment by gregtatcam** [comment]
`Details` field is used in multiple places as an inner object (`SField`), but it includes different fields. It must have a unique field name in each case, otherwise you are not going to be able to define the content of these inner objects in `InnerObjectFormats.cpp`.

**Comment by dangell7** [comment]
Is the outstanding principal the principal amount - drawable amount? This might create issues on a term loan if you allow multiple draw down events.

Doing an "Open Ended" Loan Pool makes things much more complex. Iterating through both Contributors and Composition is alot of work.
I would limit the number of contributors in a pool.

If the composition is submitted as part of the transactor, is it possible that the loan provider can choose whoever they want? Is this a legal issue? Could that loan/borrower always choose 100% of UserA assets thereby avoiding interest to specific liquidity providers in the pool? 

Updating every contributor is a lot for every deposit/loan/withdraw transaction. Not a big deal if the number is limited but at 1000 contributors thats not going to work well. Even more so, allowing withdraw/deposit of amounts "pledged" during a loan will make things very complex.

You can calculate the withdraw amount in the transactor if you have the total outstanding loans on an object but this means that the user couldn't withdraw any amount "pledged" until the loan is closed. This is how I would structure it. Allowing the User to withdraw after each payment is adding complexity that I really dont believe is needed. 

Ex. You and I deposit 100 each. A loan is made for 100. Thats 50% you and 50% me. I can withdraw 50 now and 50 + interest when the loan closes (or if the loan defaults then less). By allowing rolling withdraws you also open the door for users to each month pull the interest + principal they pledged.

Max Cover doesnt make any sense. In general the pool delegate doesn't make sense to me.
The cover or collateral should come from the borrower or a third party on behalf of the borrower which can be different for each loan.
The account that gives the collateral is not the account that receives interest. The composition is who receives interest or the pool liquidity providers. Each loan should have a collateral provider based on the PoolCollateral.min. That collateral is return at the end or taken in the event of default. Maybe I misunderstand the pool delegate.

GracePeriod & PaymentInterval are UINT32.

Some additional notes:

Loans take awhile to underwrite so by the time the loan is finalized, there may not be liquidity in the pool. Underwriters would need to manage this. You could solve this by having another transaction for LoanProposal to "lock" funds so that its clear when underwriting how much funds are available. The idea you have is that more liquidity could be deposited, which would require asking the pool or others to deposit more liquidity. 

This makes sense in theory but imo its backwards. In practice you would write a loan application, then seek liquidity or underwrite the loan and that would then receive approval from the liquidity providers.

1. Application
2. Underwriting
3. Approval

I would do this. 

`LoanPropose` - This is an agnostic object on the ledger the owner is the borrower. Its not created with a poolID. There is no lender, or pool delegate.
`ApproveLoan` - This is a voting mechanism of a pool. The votes come from the pool managers. When approved the pool id is added to the LoanObject.
`VotePoolManager` - Vote to add or remove pool managers (handles the governance structure of the pool)

- You have PoolDelegate but imo the pool liquidity providers should be the ones who vote managers in and out of the pool delegate role. In your example the PoolDelegate is whoever created the pool.

Imo this makes it much more peer to peer. The borrower can create their application and submit that to the ledger for any pool to approve of.

**Comment by Tapanito** [comment]
Closing the discussion as the specification is in a good spot. For further feedback please refer to the pull request https://github.com/XRPLF/XRPL-Standards/pull/240

**Comment by argus-forge** [comment]
Suggested Clarification:

Management Fee: This is a fee charged by the Loan Broker, calculated as a percentage of the interest earned on loans. It's deducted from the interest that would otherwise go to the Vault depositors. Essentially, borrowers pay the full interest, but before that interest reaches depositors, the Loan Broker takes their cut.

This explanation makes it clear:

That the fee is part of the interest income.

It's taken before distribution to depositors, thus affecting the yield they receive from the loans.

**Reply by Tapanito** [reply]
That's a great clarification, thank you. I'll update the spec. Though the discussion is closed, let's move this conversation to the PR :bow:

**Reply by argus-forge** [reply]
You're welcome! And my apologies. Only my second time commenting on Github; still learning. Thanks! Here to help!

**Comment by Cbuske** [comment]
Is this lending protocol suppose to work in tandem with traditional regulatory guidelines or are you creating a defi lending protocol to compete in the traditional lending world? I am curious as I am a mortgage expert and I have been developing something that could potentially revolutionize how mortgages are secured streamlining the entire process for borrowers. I would like to know if this lending protocol is something that I can incorporate into my off-chain application and underwriting system that actually connects to traditional secondary investors. This protocol seems very interesting, but how practical is in when it comes to abiding by current lending regulations or guidelines?  I am interested in collaborating and exploring this protocol more in detail. I have a vision to create something that is more applicable to mainstream adoption in the industry that correlates to all lending laws and regulations.

**Reply by Tapanito** [reply]
Let's explore how we could collaborate. Could you email me at vtumas@ripple.com, and I will connect you with the right people?

**Comment by Cbuske** [comment]
Hello Vito,
Thank you for reaching out via email. I would love to explore how we could collaborate to develop a new system that could transform the entire mortgage industry for the better. I will email you from my business account coming from ***@***.*** I greatly appreciate the introduction. 
Best Regards,
Craig Buske
    On Friday, January 31, 2025 at 07:11:13 AM CST, Vito Tumas ***@***.***> wrote:  
 
 


Let's explore how we could collaborate. Could you email me at ***@***.***, and I will connect you with the right people?

**Comment by Cbuske** [comment]
Hello Vito,
I am following up with you based on our last email thread. I am just checking to see if you received my email from my Mortgage Leaders business account regarding collaborating on an collateral mortgage that involves combining AI/ML technology and your lending protocol. I only see this type of system feasible on the XRPL since Ripple already has the regulations for money transmissions in place through the NMLS. Please let me know. I look forward hearing from you. 
Best Regards,
Craig BuskeMortgage Leaders, LLC.SVP of Mortgage Lending
    On Friday, January 31, 2025 at 07:11:13 AM CST, Vito Tumas ***@***.***> wrote:  
 
 


Let's explore how we could collaborate. Could you email me at ***@***.***, and I will connect you with the right people?
